Wabstic import

gh-PonyM opened the pull request:
- Improved errors when parsing major and proporz elections.
- Switching expected headers in wabstic import scripts from `gewahlt` to `gewaehlt`. Fixed failing import of majorz election test dataset that raised "Missing column: gewahlt".
- Import of new majorz and proporz test data was successfull

Commit:
Adds modifications from archive search branch

- Passing tests
- Removing archive search specific modifications

Commit:
Changes expected column title according to csv files (gewaehlt)

Commit:
Refactoring validation functions

Commit:
Fixes use of entity_id before possible assignment (in try block)

- Move into the else block, that gets executed if try does not raise
any error

Commit:
Adds helper function

Commit:
Adds annotations and debug prints

Commit:
Refactoring

Commit:
Refactoring

Commit:
Sorts printed errors to compare to test

Commit:
Refactoring

- test catches first of execpetions, the rest works the same way

Commit:
Renames gewahlt to gewaehlt, refactoring, improved errors

Commit:
Fixes '' in csv to be 0 (integer) according to CandidateResult.votes

Commit:
Refactoring

Commit:
Improves error msgs for kandidatengde, fixes tests

Commit:
Refactoring, line_is_relevant can be shared for proporz and majorz

Commit:
Linting

Commit:
Refactoring

Commit:
Reduces redundancy of checking for missing columns

href commented on onegov/election_day/formats/common.py:
```diff
@@ -142,3 +142,44 @@ def load_csv(
+def line_is_relevant(line, number, district=None):
+    if district:
+        return line.sortwahlkreis == district and line.sortgeschaeft == number
+    else:
```
You can remove the else, since the if above always returns:

```python
if district:
    return line.sortwahlkreis == district and line.sortgeschaeft == number

return line.sortgeschaeft == number
```

href commented on onegov/election_day/formats/common.py:
```diff
@@ -142,3 +142,44 @@ def load_csv(
+    :param none_be_zero: raises ValueError if line.col is None
+    :return: integer value of line.col
+    """
+    assert hasattr(line, col), 'Check done in load_csv'
```
What does 'Check done in load_csv' mean? Could you elaborate here?

If you need longer texts, you can also write assertions like this:

```python
assert hasattr(line, col), f"""
     {col} was not found on {line} - this should have failed in load_csv
"""

gh-PonyM replied:
I forgot to drop this line, since in load_csv in common.py, checks if the expected columns are there is performed already. The comment was a remainder of that.

href commented on onegov/election_day/formats/common.py:
```diff
@@ -142,3 +142,44 @@ def load_csv(
+        return line.sortgeschaeft == number
+
+
+def validate_integer(line, col, none_be_zero=True):
```
I'd say `non_be_zero` should be named `treat_none_as_zero`. What do you think?

gh-PonyM replied:
it's better like this, I also added a default kwarg to return that default and changed the lines where the votes are received for the candidate.

href commented on onegov/election_day/formats/election/wabstic_majorz.py:
```diff
@@ -46,15 +47,16 @@
-    entity_id = int(line.bfsnrgemeinde or 0)
+def get_entity_id(line):
+    col = 'bfsnrgemeinde'
+    # try:
```
Commented out code should not be committed, unless it is meant to be activated in the near future (for example, if a Puppet deployment has been prepared and tested, but not yet rolled out).

href commented on onegov/election_day/formats/election/wabstic_majorz.py:
```diff
@@ -79,6 +81,16 @@ def import_election_wabstic_majorz(
     election_id = election.id
 
+    def has_no_lines(lines, filename):
+        if not list(lines):
```
You consume the whole list of lines, to check if there's at least one line. That seems wasteful. If lines is not a generator, you can just check for by running `if lines`.

If lines is a generator, don't have to consume it to check for it. You can do either this:
```python
for line in lines:
     return False  # has lines

return True # has no lines
```
Or by getting the next value of the generator with a fallback:
```python
if next(lines, None):
    return True # has lines

return False  # has no lines

gh-PonyM replied:
Yes, it's definitively a blocker...

href commented on onegov/election_day/formats/election/wabstic_proporz.py:
```diff
@@ -179,6 +185,16 @@ def import_election_wabstic_proporz(
     election_id = election.id
 
+    def has_no_lines(lines, filename):
+        if not list(lines):
```
Same as above, this should not traverse all lines.

href reviewed (changes requested):
Looks good to me, aside from a few small things. You certainly want to avoid consuming all the CSV lines however - that's a blocker for me.

I think coercing `None` to `0` makes sense, as `CandiateResult.votes` wants it that way. When in doubt we probably want to stick with the existing model. A lot more thinking goes into building up the model than adding features later, so I guess we trust that it makes sense as is 😉

Commit:
Removes validation since its done in csv import

Commit:
removes validation of empty lines

- unclear which files could be empty and must be importable
- does not fit script flow causing errors anyway and aborting import

Commit:
Refactor

Commit:
Maintains old return value 'None' if stimmen is '' is csv file

Commit:
Linting
